compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java

```diff
@@ -3172,6 +3180,127 @@ private STNode parseActionOrExpressionInLhs(SyntaxKind nextTokenKind, STNode lhs
         return parseExpressionRhs(nextTokenKind, DEFAULT_OP_PRECEDENCE, lhsExpr, false, true);
     }
 
+    /**
+     * <p>
+     * Parse a new expression.
+     * </p>
+     * <code>
+     *  new-expr := explicit-new-expr | implicit-new-expr
+     *  explicit-new-expr := new type-descriptor ( arg-list )
+     *  implicit-new-expr := new [( arg-list )]
+     * </code>
+     *
+     * @return Parsed NewExpression node.
+     */
+    private STNode parseNewExpression() {
+        startContext(ParserRuleContext.NEW_EXPRESSION);
+        STNode newKeyword = parseNewKeyword();
+        STNode newExpression = parseNewExpression(newKeyword);
+        endContext();
+        return newExpression;
+    }
+
+    /**
+     * <p>
+     * Parse `new` keyword.
+     * </p>
+     *
+     * @return Parsed NEW_KEYWORD Token.
+     */
+    private STNode parseNewKeyword() {
+        STToken token = peek();
+        if (token.kind == SyntaxKind.NEW_KEYWORD) {
+            return consume();
+        } else {
+            Solution sol = recover(token, ParserRuleContext.NEW_KEYWORD);
+            return sol.recoveredNode;
+        }
+    }
+
+    private STNode parseNewExpression(STNode newKeyword) {
+        STNode token = peek();
+        return parseNewExpression(token.kind, newKeyword);
+    }
+
+    /**
+     * <p>
+     * Parse an implicit or explicit expression.
+     * </p>
+     * @param kind next token kind.
+     * @param newKeyword parsed node for `new` keyword.
+     * @return Parsed new-expression node.
+     */
+    private STNode parseNewExpression(SyntaxKind kind, STNode newKeyword) {
+        switch (kind) {
+            case OPEN_PAREN_TOKEN:
+                return parseImplicitNewExpression(newKeyword);
+            case SEMICOLON_TOKEN:
+                return STNodeFactory.createImplicitNewExpression(newKeyword, STNodeFactory.createEmptyNode());
+            case IDENTIFIER_TOKEN:
+            case OBJECT_KEYWORD:
+                // TODO: Support `stream` keyword once introduced
+                return parseExplicitNewExpression(newKeyword);
+            default:
+                Solution sol = recover(peek(), ParserRuleContext.NEW_EXPRESSION, newKeyword);
+                return parseNewExpression(sol.recoveredNode.kind, newKeyword);
+        }
+    }
+
+    /**
+     * <p>
+     * Parse an Explicit New expression.
+     * </p>
+     * <code>
+     *  explicit-new-expr := new type-descriptor ( arg-list )
+     * </code>
+     *
+     * @param newKeyword Parsed `new` keyword.
+     * @return the Parsed Explicit New Expression.
+     */
+    private STNode parseExplicitNewExpression(STNode newKeyword) {
+        startContext(ParserRuleContext.EXPLICIT_NEW_RHS);
+        STNode typeDescriptor = parseTypeDescriptor();
```

Comment:
Need to validate whether this is a object-type-desc or a type-reference

Reply:
Since I come to this stage with only OBJECT_KEYWORD i didn't think i need to do that. But since the discussion we had to enable all the types and then validate, to give a meaningful error, I think i have to change the logic here